nautobot/dcim/models/locations.py

```diff
@@ -251,6 +269,32 @@ class Meta:
     def __str__(self):
         return self.name
 
+    @classmethod
+    def get_natural_key_fields(cls):
+        natural_key = []
+        name = "name"
+        for _ in range(cls.objects.max_tree_depth() + 1):
+            natural_key.append(name)
+            name = f"parent__{name}"
+        return natural_key
+
+    def natural_key(self):
+        """
+        Custom natural key implementation for Location.
+
+        Needed because with the uniqueness constraint of (name, parent) we end up with a theoretically infinite
+        recursive natural key.
+        """
+        try:
+            return [ancestor.name for ancestor in reversed(self.ancestors(include_self=True))]
+        except Location.DoesNotExist:
```

Comment:
I ran into a case (IIRC it was during dumpdata/loaddata) where `self.ancestors()` was returning objects that threw a DoesNotExist exception, hence this fallback logic.